src/lib/nested_property.js

```diff
@@ -113,24 +115,36 @@ function npGet(cont, parts) {
 }
 
 /*
- * Check known non-data-array arrays (containers). Data arrays only contain scalars,
- * so parts[end] values, such as -1 or n, indicate we are not dealing with a dataArray.
- * The ONLY case we are looking for is where the entire array is selected, parts[end] === 'x'
- * AND the replacement value is an array.
+ * Can this value be deleted? We can delete any empty object (null, undefined, [], {})
+ * EXCEPT empty data arrays. Info arrays can be safely deleted, but not deleting them
+ * has no ill effects other than leaving a trace or layout object with some cruft in it.
+ * But deleting data arrays can change the meaning of the object, as `[]` means there is
+ * data for this attribute, it's just empty right now while `undefined` means the data
+ * should be filled in with defaults to match other data arrays. So we do some simple
+ * tests here to find known non-data arrays but don't worry too much about not deleting
+ * some arrays that would actually be safe to delete.
  */
-function isDataArray(val, key) {
-
-    var containers = ['annotations', 'shapes', 'range', 'domain', 'buttons'],
-        isNotAContainer = containers.indexOf(key) === -1;
-
-    return isArray(val) && isNotAContainer;
+var INFO_PATTERNS = /(^|.)((domain|range)(\.[xy])?|args|parallels)$/;
```

Comment:
Ok, well this is the only part I dislike about this PR. I'm ok with leaving as is for now, but should think about ways to :hocho: this down the road.

Solution 1: simply don't delete `info_array` attributes, like `data_array`

Solution 2: use the plot schema to look up all `info_array` attributes. This doesn't have to be done inside nested property where objects are set w/o a reference to their parent object (if any). Maybe this could be done only during `restyle`, `relayout` and `update`.

Solution 3: something else